ADR 007: Solo machine signature construction (#1865)
cosmos/ibc-go: docs/architecture/adr-007-solomachine-signbytes.md

# ADR 007: Solo machine sign bytes

## Changelog
* 2022-08-02: Initial draft

## Status

Accepted, applied in v6

## Context

> The solo machine implemention up until ibc-go v6 constructed sign bytes using a `DataType` which described what type of data was being signed. 

Reviewer: ```suggestion
The `06-solomachine` implemention up until ibc-go v6 constructed sign bytes using a `DataType` which described what type of data was being signed. 
```

> The `06-solomachine` implemention up until ibc-go v6 constructed sign bytes using a `DataType` which described what type of data was being signed. 

Reviewer 2: nit
```suggestion
The `06-solomachine` implemention up until `ibc-go` v6 constructed sign bytes using a `DataType` which described what type of data was being signed. 
```

> The `06-solomachine` implemention up until `ibc-go` v6 constructed sign bytes using a `DataType` which described what type of data was being signed. 

Reviewer 3: ```suggestion
The `06-solomachine` implemention up until ibc-go v6 constructed sign bytes using a `DataType` which described what type of data was being signed. 
```

This design decision arose from a misunderstanding of the security implications.
It was noted that the proto definitions do not [provide uniqueness](https://github.com/cosmos/cosmos-sdk/pull/7237#discussion_r484264573) which is a necessity for ensuring two signatures over different data types can never be the same.
What was missed is that the uniqueness is not provided by the proto definition, but by the usage of the proto definition. 
The path provided by core IBC will be unique and is already encoded into the signature data.

> Thus two different paths with the same data values will encode differently which provides signatuer uniqueness. 

Reviewer 4: ```suggestion
Thus two different paths with the same data values will encode differently which provides signature uniqueness. 
```

> Furthermore,  the current construction does not support the proposed changes in the spec repo to support [Generic Verification functions](https://github.com/cosmos/ibc/issues/684).

Reviewer: ```suggestion
Furthermore, the current construction does not support the proposed changes in the spec repo to support [Generic Verification functions](https://github.com/cosmos/ibc/issues/684).
```
